File: src/ros_tutorials-noetic-devel/turtleBot3/ros_turtlebot_control/src/poseGoal.py
#!/usr/bin/env python
import rospy
#import utils.geo_maths as geo_maths
from std_msgs.msg import Float32
from geometry_msgs.msg import TwistStamped, Twist, Pose
from gazebo_msgs.msg import ModelState
from gazebo_msgs.srv import SetModelState
from ros_turtlebot_control.srv import *
from nav_msgs.msg import Odometry

pose_x=Float32();
pose_y=Float32();
theta=Float32();
def callback(data):
    pose_x.data=data.pose.pose.position.x;
    pose_y.data=data.pose.pose.position.y;
    theta.data=0;

def main():
    rospy.init_node('set_vel_py')
    rate=rospy.Rate(5)
    rospy.Subscriber('/ground_truth/state', Odometry, callback)
    while not rospy.is_shutdown():  
       x=pose_x.data;
       y=pose_y.data;
       the_ta=0;
       moveToPose=MoveToPoseRequest()
       moveToPose.x=x;
       moveToPose.y=y;
       moveToPose.theta=the_ta;
       rate.sleep()
       rospy.wait_for_service('/turtle/move_to_pose')
       try:
           set_state = rospy.ServiceProxy('/turtle/move_to_pose', MoveToPose)
           resp = set_state(moveToPose)
       except rospy.ServiceException, e:
          print "Service call failed: %s" % e

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: src/ros_tutorials-noetic-devel/turtleBot3/ros_turtlebot_control/src/poseGoal.cpp
#include <ros/ros.h>
#include <gazebo_msgs/SetModelState.h>
#include <cstdlib>
#include <std_msgs/Float32.h>
#include <geometry_msgs/TwistStamped.h>
#include <geometry_msgs/Pose.h>
#include <geometry_msgs/Twist.h>
#include <ros_turtlebot_control/MoveToPose.h>
#include <nav_msgs/Odometry.h>
//I'm going to implement user-specified twist values later.
  std_msgs::Float32 pose_x;
  std_msgs::Float32 pose_y;
  std_msgs::Float32 theta;
  std_msgs::Float32 ori_x;
  std_msgs::Float32 ori_y;
  std_msgs::Float32 ori_z;
  std_msgs::Float32 ori_w;


void callback(const nav_msgs::Odometry::ConstPtr& msg)
{
  pose_x.data=msg->pose.pose.position.x;
  pose_y.data=msg->pose.pose.position.y;
  theta.data=0;
  ori_x.data=msg->pose.pose.orientation.x;
  ori_y.data=msg->pose.pose.orientation.y;
  ori_z.data=msg->pose.pose.orientation.z;
  ori_w.data=msg->pose.pose.orientation.w;

}


int main (int argc, char **argv)
{
  ros::init(argc,argv,"turtlebot3_waffle_pi");
  ros::NodeHandle n;
  ros::Rate loop_rate(1);
  ros::Subscriber sub = n.subscribe("/ground_truth/state", 1, callback);


  while (ros::ok()){
     ros::spinOnce();
     double x=pose_x.data;
     double y=pose_y.data;
     double the_ta=theta.data;
     ros::ServiceClient client = n.serviceClient<ros_turtlebot_control::MoveToPose>("/turtle/move_to_pose");
     ros_turtlebot_control::MoveToPose moveToPose;
     moveToPose.request.x = x;
     moveToPose.request.y = y;
     moveToPose.request.theta = the_ta;
     loop_rate.sleep();
  if (client.call(moveToPose))
  {
    ROS_INFO("BRILLIANT!!!");
    ROS_INFO("%f, %f",pose_x.data, pose_y.data);
  }  
  else
  {
    ROS_ERROR("Failed to call service ");
    return 1;
  }

  }




//  ros::spin();
  return 0;  
}
